crates/biome_js_analyze/src/semantic_analyzers/nursery/no_unused_imports.rs

```diff
@@ -0,0 +1,223 @@
+use crate::{semantic_services::Semantic, JsRuleAction};
+use biome_analyze::{context::RuleContext, declare_rule, ActionCategory, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_diagnostics::Applicability;
+use biome_js_factory::make;
+use biome_js_semantic::ReferencesExtensions;
+use biome_js_syntax::{
+    binding_ext::AnyJsBindingDeclaration, AnyJsImportClause, JsFileSource, JsIdentifierBinding,
+    JsImport, JsImportNamedClause, JsLanguage, JsNamedImportSpecifierList, JsSyntaxNode, T,
+};
+use biome_rowan::{
+    AstNode, AstSeparatedList, BatchMutation, BatchMutationExt, NodeOrToken, SyntaxResult,
+};
+
+declare_rule! {
+    /// Disallow unused imports.
+    ///
+    /// Unused imports usually are result of incomplete refactoring.
+    ///
+    /// There is one exception to the rule: the `React` import.
+    /// Importing the `React` variable was a mandatory pattern until some time ago:
+    /// For the time being this rule will ignore it,
+    /// but this **might change in the future releases**.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// // Comment
+    /// import A from 'mod';
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// // Comment
+    /// import * as A from 'mod';
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// import { type A, B } from 'mod';
+    ///
+    /// export { B }
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```ts
+    /// import { A, type B } from 'mod';
+    ///
+    /// function f(arg: B): A {
+    ///     return new A(arg);
+    /// }
+    /// ```
+    ///
+    /// ```jsx
+    /// import React from 'react';
+    ///
+    /// function foo() {
+    ///     return <div />;
+    /// };
+    ///
+    /// foo();
+    /// ```
+    pub(crate) NoUnusedImports {
+        version: "next",
+        name: "noUnusedImports",
+        recommended: false,
+    }
+}
+
+impl Rule for NoUnusedImports {
+    type Query = Semantic<JsIdentifierBinding>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let binding = ctx.query();
+        if !is_import(&binding.declaration()?) {
+            return None;
+        }
+
+        if ctx.source_type::<JsFileSource>().variant().is_jsx() {
+            let name = binding.name_token().ok()?;
+            let name = name.text_trimmed();
+            // Old code import `React` but only use it indirectly after transpiling JSX.
```

Comment:
I don't really understand the comment...

Reply:
It is a copy-paste from the comment of `noUnusedVariables`.